internal/app/app.go

```diff
@@ -1,27 +1,140 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
 package app
 
 import (
 	"context"
+	"time"
+
+	"github.com/gardener/etcd-wrapper/internal/types"
 
+	"github.com/gardener/etcd-wrapper/internal/bootstrap"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/server/v3/embed"
 	"go.uber.org/zap"
 )
 
 // Application is a top level struct which serves as an entry point for this application.
 type Application struct {
-	ctx        context.Context
-	logger     *zap.Logger
-	caCertPath string
+	ctx              context.Context
+	cancelFn         context.CancelFunc
+	etcdInitializer  bootstrap.EtcdInitializer
+	cfg              *embed.Config
+	etcdClient       *clientv3.Client
+	etcd             *embed.Etcd
+	waitReadyTimeout time.Duration
+	logger           *zap.Logger
+	etcdReady        bool // should have only one actor that updates it, queryAndUpdateEtcdReadiness()
 }
 
-// Setup sets up the application.
+// NewApplication initializes and returns an application struct
+func NewApplication(ctx context.Context, cancelFn context.CancelFunc, brConfig *types.BackupRestoreConfig, waitReadyTimeout time.Duration, logger *zap.Logger) (*Application, error) {
+	etcdInitializer, err := bootstrap.NewEtcdInitializer(brConfig, logger)
+	if err != nil {
+		return nil, err
+	}
+	return &Application{
+		ctx:              ctx,
+		cancelFn:         cancelFn,
+		etcdInitializer:  etcdInitializer,
+		waitReadyTimeout: waitReadyTimeout,
+		logger:           logger,
+	}, nil
+}
+
+// Setup sets up etcd by triggering initialization of the etcd DB.
 func (a *Application) Setup() error {
-	//sets up etcd by calling bootstrap.InitializeEtcd()
+	// Set up etcd
+	cfg, err := a.etcdInitializer.Run(a.ctx)
+	if err != nil {
+		return err
+	}
+	a.cfg = cfg
 	return nil
 }
 
-// Start starts this application.
+// Start sets up readiness probe and starts an embedded etcd.
 func (a *Application) Start() error {
+	var err error
+
+	// Create etcd client for readiness probe
+	cli, err := a.createEtcdClient()
+	if err != nil {
+		return err
+	}
+	a.etcdClient = cli
+	defer a.Close()
+
+	// Setup readiness probe
 	go a.SetupReadinessProbe()
+
 	// Create embedded etcd and start.
+	if err = a.startEtcd(); err != nil {
+		return err
+	}
+	// Delete validation marker after etcd starts successfully
+	if err = bootstrap.CleanupExitCode(types.DefaultExitCodeFilePath); err != nil {
+		a.logger.Warn("failed to clean-up last captured exit code", zap.Error(err))
+	}
```

Comment:
AFAIU, you are starting an embedded etcd by calling `startEtcd()` then you are removing exit code file of previous exit, Is it ?

Reply:
Correct. Previous exit code file is removed only when etcd is started